drivers/sensor/mb7040/mb7040.c

```diff
@@ -0,0 +1,216 @@
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/gpio.h>
+
+#define RANGE_CMD              0x51
+#define MB7040_HAS_STATUS_GPIO DT_ANY_INST_HAS_PROP_STATUS_OKAY(status_gpios)
+
+LOG_MODULE_REGISTER(mb7040, CONFIG_SENSOR_LOG_LEVEL);
+
+struct mb7040_data {
+	uint16_t distance_cm;
+	struct k_sem read_sem;
+#if MB7040_HAS_STATUS_GPIO
+	struct gpio_callback gpio_cb;
+#endif
+};
+
+struct mb7040_config {
+	struct i2c_dt_spec i2c;
+#if MB7040_HAS_STATUS_GPIO
+	struct gpio_dt_spec status_gpio;
+#endif
+};
+
+#if MB7040_HAS_STATUS_GPIO
+static void status_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct mb7040_data *data = CONTAINER_OF(cb, struct mb7040_data, gpio_cb);
+
+	k_sem_give(&data->read_sem);
+}
+#endif
+
+static int mb7040_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct mb7040_config *cfg = (struct mb7040_config *)dev->config;
+	struct mb7040_data *data = (struct mb7040_data *)dev->data;
+	uint8_t cmd = RANGE_CMD;
+	int ret;
+	uint8_t read_data[2];
+
+	if (chan != SENSOR_CHAN_DISTANCE && chan != SENSOR_CHAN_ALL) {
+		LOG_ERR("Sensor only supports distance");
+		return -EINVAL;
+	}
+
+	k_sem_reset(&data->read_sem);
+
+#if MB7040_HAS_STATUS_GPIO
+	/* Check if status_gpio port is present */
+	if (cfg->status_gpio.port != NULL) {
+		/* Enable interrupt before writing */
+		ret = gpio_pin_interrupt_configure_dt(&cfg->status_gpio, GPIO_INT_EDGE_FALLING);
+		if (ret != 0) {
+			LOG_ERR("Failed to configure interrupt: %d", ret);
+			return ret;
+		}
+	}
+#endif
+
+	/* Write range command to sensor */
+	ret = i2c_write_dt(&cfg->i2c, &cmd, 1);
+	if (ret != 0) {
+		LOG_ERR("I2C write failed with error %d", ret);
+#if MB7040_HAS_STATUS_GPIO
+		/* Disable interrupt before returning error*/
+		if (cfg->status_gpio.port != NULL) {
+			gpio_pin_interrupt_configure_dt(&cfg->status_gpio, GPIO_INT_DISABLE);
+		}
+#endif
+		return ret;
+	}
+
+	ret = k_sem_take(&data->read_sem, K_MSEC(CONFIG_MB7040_DELAY_MS));
+
+#if MB7040_HAS_STATUS_GPIO
+	/* Check if status_gpio port is present */
+	if (cfg->status_gpio.port != NULL) {
+		gpio_pin_interrupt_configure_dt(&cfg->status_gpio, GPIO_INT_DISABLE);
+		if (ret == 0) {
+			/* Success: interrupt fired and semaphore taken */
+		} else if (ret == -EAGAIN) {
+			LOG_ERR("Semaphore take timed out");
+			return -EAGAIN;
+		} else {
+			return ret;
+		}
+	} else {
+		if (ret == 0) {
+			/* Should not happen: semaphore only given by GPIO interrupt */
+			LOG_ERR("Semaphore taken unexpectedly without status GPIO");
+			return 0;
+		} else if (ret == -EAGAIN) {
+			/* Timeout is expected: treat as success */
+		} else {
+			return ret;
+		}
+	}
+#else
+	if (ret == 0) {
+		/* Should not happen: semaphore only given by GPIO interrupt */
+		LOG_ERR("Semaphore taken unexpectedly without status GPIO");
+		return 0;
+	} else if (ret == -EAGAIN) {
+		/* Timeout is expected: treat as success */
+	} else {
+		return ret;
+	}
```

Comment:
I think trying to use the semaphore even with no GPIO interrupt is making this driver logic unnecessary complicated.

I suggest only using the semaphore when STATUS GPIO is defined, otherwise just use a delay.

Reply:
The sensor stops responding to I2C requests for the duration of measurement. So when the status GPIO is not available polling (like an AT24 EPROM after a write command) is a way to get data faster than a fixed delay. Is this a viable alternative to a fixed delay?